### PR TITLE
[Feature] infer property for mv

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalScanOperator.java [hunk 1]
```diff
@@ -63,6 +63,14 @@ public List<ColumnRefOperator> getOutputColumns() {
         return outputColumns;
     }
 
+    public List<ColumnRefOperator> getRealOutputColumns() {
+        if (projection != null) {
+            return projection.getOutputColumns();
+        } else {
+            return outputColumns;
+        }
+    }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalScanOperator.java:72]: why don't modify outputColumns？

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 2]
```diff
@@ -1246,28 +1261,37 @@ private PlanFragment removeExchangeNodeForLocalShuffleAgg(PlanFragment inputFrag
...
         }
 
-        @Override
-        public PlanFragment visitPhysicalHashAggregate(OptExpression optExpr, ExecPlan context) {
-            PhysicalHashAggregateOperator node = (PhysicalHashAggregateOperator) optExpr.getOp();
-            PlanFragment originalInputFragment = visit(optExpr.inputAt(0), context);
-
-            PlanFragment inputFragment = removeExchangeNodeForLocalShuffleAgg(originalInputFragment, context);
-            boolean withLocalShuffle = inputFragment != originalInputFragment;
+        static class AggregateExprInfo {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1264]: private

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Infer key for each operator from the query plan
+ */
+public class KeyInference extends OptExpressionVisitor<KeyInference.KeyPropertySet, Void> {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java:37]: what's Key？identity？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MVOperatorProperty.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.rule.mv;
+
+public class MVOperatorProperty {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MVOperatorProperty.java:5]:
```suggestion
public class MVPlanProperty {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+    public static class ModifyOp {
+        private final EnumSet<ModifyKind> modifySet;
+        private final EnumSet<UpdateKind> updateSet;
+
+        public static final ModifyOp NONE = new ModifyOp(ModifyKind.NONE, UpdateKind.NONE);
+        public static final ModifyOp INSERT_ONLY = new ModifyOp(ModifyKind.INSERT_ONLY, UpdateKind.NONE);
+        public static final ModifyOp UPSERT = new ModifyOp(ModifyKind.UPSERT, UpdateKind.ONLY_AFTER);
+        public static final ModifyOp ALL = new ModifyOp(ModifyKind.ALL, UpdateKind.ALL);
+
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java:113]: so What's the ModifyOp ? Different ModifyOps will generate which physical property?  

Is there some information to tell Physical operator which can not generate `retract` messages?

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 6]
```diff
@@ -2543,5 +2425,305 @@ public PlanFragment visitPhysicalCTEAnchor(OptExpression optExpression, ExecPlan
...
+            JoinNode.DistributionMode distributionMode =
+                    inferDistributionMode(optExpr, leftFragmentPlanRoot, rightFragmentPlanRoot);
+            if (distributionMode.equals(JoinNode.DistributionMode.REPLICATED)) {
+                throw new NotImplementedException("Replicated join is not supported");
+            }
+            // 2. Build join expression
+            JoinExprInfo joinExpr = buildJoinExpr(optExpr, context);
+            List<Expr> eqJoinConjuncts = joinExpr.eqJoinConjuncts;
+            List<Expr> otherJoinConjuncts = joinExpr.otherJoin;
+            List<Expr> conjuncts = joinExpr.conjuncts;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:2522]: Maybe we can add more restrictions right now, StreamJoin cannot support un-equal joins?

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by dulong41** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalScanOperator.java (hunk 1 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalScanOperator.java:72]: I modified it in another PR, and many test cases failed.

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java (hunk 2 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1267]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    public ModifyOp visitPhysicalStreamAgg(OptExpression optExpression, Void ctx) {
+        ModifyOp inputModify = infer(optExpression.inputAt(0));
+        PhysicalStreamAggOperator agg = (PhysicalStreamAggOperator) optExpression.getOp();
+
+        // TODO(murphy) optimize for non-retractable scenario
+        ModifyOp res;
+        if (inputModify.isInsertOnly()) {
+            res = ModifyOp.UPSERT;
+        } else {
+            res = ModifyOp.ALL;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java:99]: - `ModifyKind` include `INSERT`, `UPDATE`, `DELETE`, and they could be combined as `INSERT_ONLY`, `ALL`, `UPSERT`

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        } else {
+            throw new NotImplementedException("StreamAggregation without group by is not supported");
+        }
+
+        return agg.getKeyPropertySet();
+    }
+
+    public static class KeyProperty {
+        public final boolean unique;
+        public final ColumnRefSet columns;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java:195]: Each operator could inference multiple `KeyProperty`, so it's stored as `KeyPropertySet`.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        }
+        res.addKey(KeyProperty.of(projectColumns, false));
+
+        return res;
+    }
+
+    private KeyPropertySet visitTable(OlapTable olapTable, PhysicalOperator scan, Projection projection,
+                                      List<ColumnRefOperator> outputColumns,
+                                      Map<ColumnRefOperator, Column> columnMap) {
+        List<ColumnRefOperator> keyColumnRefs;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java:79]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            keyColumnRefs = projection.getOutputColumns();
+        }
+
+        // If the table contains unique key and output the whole unique-keys
+        List<Column> tableKeyColumns = olapTable.getKeyColumns();
+        Set<Column> keyRefs = keyColumnRefs.stream().map(columnMap::get).collect(Collectors.toSet());
+        boolean unique = olapTable.getKeysType().equals(KeysType.PRIMARY_KEYS) &&
+                keyRefs.containsAll(tableKeyColumns);
+
+        KeyProperty key = KeyProperty.of(new ColumnRefSet(keyColumnRefs), unique);
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/KeyInference.java:93]: We need to support the duplicate key table, which is not unique. For join two duplicate table, we need to define the key of sink table, which could be inferred through this routine.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    // TODO(murphy) read from user property, support custom the ModifyOp behavior
+    private ModifyOp visitOlapTable(Table table) {
+        if (!table.isOlapTable()) {
+            throw new NotImplementedException("Only OLAP table is supported");
+        }
+        OlapTable olapTable = (OlapTable) table;
+        ModifyOp res;
+        if (olapTable.getKeysType().equals(KeysType.DUP_KEYS)) {
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java:52]: Not considered right now.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java (hunk 5 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java:113]: Each stream operator has a `ModifyOp`, such as `INSERT_ONLY`, `UPSERT`, `ALL`.

Take the `StreamAgg` Operator as example:
- If its input is `INSERT_ONLY`, the agg itself is `UPSERT`, so it generates `INSERT` and `DELETE` only, but not `UPDATE`
- If its input is `ALL`, the agg would output all kinds of `ModifyOp`

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java (hunk 6 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:2525]: It should be checked in `Analyzer`, not here

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by Seaven — DISMISSED** [review]
LGTM，try to reslove code smell plz

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamJoinOperator.java [hunk 12]
```diff
@@ -35,7 +60,23 @@ public String toString() {
...
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
+        PhysicalStreamJoinOperator that = (PhysicalStreamJoinOperator) o;
+        return joinType == that.joinType && Objects.equals(onPredicate, that.onPredicate) &&
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamJoinOperator.java:74]: why need to consider joinHint?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+            res = ModifyOp.INSERT_ONLY;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java:53]: we could delete the DUP_KEYS table.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java [hunk 14]
```diff
@@ -26,7 +26,7 @@ public void executeTasks(TaskContext context) {
         long timeout = context.getOptimizerContext().getSessionVariable().getOptimizerExecuteTimeout();
         Stopwatch watch = context.getOptimizerContext().getTraceInfo().getStopwatch();
         while (!tasks.empty()) {
-            if (watch.elapsed(TimeUnit.MILLISECONDS) > timeout) {
+            if (timeout > 0 && watch.elapsed(TimeUnit.MILLISECONDS) > timeout) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java:29]: if we want to valid the user config, we should check it when set, not here.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java (hunk 13 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/ModifyInference.java:53]: That's a logical delete, which could not handled by mv right now. The stream operator could only accept the physical delete

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java (hunk 14 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java:29]: Not a check, here I treat the `timeout =0` as not ignore the timeout

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamJoinOperator.java (hunk 12 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamJoinOperator.java]: fixed

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Review by Seaven — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run all

**Comment by mofeiatwork** [comment]
run all

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 402 / 507 (79.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalProjectOperator.java|0|1|00.00%|[75]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/ColumnRefSet.java|0|1|00.00%|[70]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpression.java|0|4|00.00%|[142, 143, 144, 145]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamScanOperator.java|0|21|00.00%|[32, 33, 34, 35, 36, 37, 39, 41, 44, 48, 52, 72, 73, 75, 76, 78, 79, 81, 82, 83, 88]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|0|1|00.00%|[240]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MVOperatorProperty.java|7|12|58.33%|[12, 13, 14, 15, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/KeyInference.java|80|114|70.18%|[50, 55, 60, 61, 62, 64, 65, 66, 67, 69, 70, 72, 102, 103, 104, 105, 107, 108, 109, 110, 120, 214, 215, 217, 218, 220, 221, 226, 239, 256, 257, 258, 269, 270]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/ModifyInference.java|58|71|81.69%|[32, 37, 42, 48, 68, 69, 70, 71, 82, 124, 141, 149, 154]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamJoinOperator.java|17|20|85.00%|[65, 68, 71]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|8|9|88.89%|[415]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/stream/PhysicalStreamOperator.java|9|10|90.00%|[34]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|206|226|91.15%|[2496, 2521, 2562, 2573, 2575, 2577, 2578, 2581, 2582, 2583, 2584, 2597, 2598, 2661, 2711, 2712, 2739, 2740, 2744, 2745]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/JoinHelper.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamJoinNode.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalScanOperator.java|3|3|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**